src/firmware_new/boards/frankie_v1/io/proto_multicast_communication.c

```diff
@@ -0,0 +1,230 @@
+                             NETCONN_JOIN);
+
+    // this buffer is used to hold serialized proto
+    uint8_t buffer[comm_profile->message_max_size];
+
+    // network buffer
+    struct netbuf* tx_buf = netbuf_new();
+
+    for (;;)
+    {
+        pb_ostream_t stream = pb_ostream_from_buffer(buffer, sizeof(buffer));
+
+        // block until protobuf has been updated
+        io_proto_multicast_communication_profile_blockUntilEvents(comm_profile,
+                                                                  PROTO_UPDATED);
+
+        // serialize proto into buffer
+        io_proto_multicast_communication_profile_acquireLock(comm_profile);
+        pb_encode(&stream, comm_profile->message_fields, comm_profile->protobuf_struct);
+        io_proto_multicast_communication_profile_releaseLock(comm_profile);
+
+        netbuf_alloc(tx_buf, stream.bytes_written);
+
+        // package payload and send over udp
+        tx_buf->p->payload = buffer;
+        netconn_sendto(conn, tx_buf, &comm_profile->multicast_address,
+                       comm_profile->port);
+    }
+
+    netbuf_delete(tx_buf);
+}
+
+void io_proto_multicast_listener_Task(void* arg)
+{
+    ProtoMulticastCommunicationProfile_t* comm_profile =
+        (ProtoMulticastCommunicationProfile_t*)arg;
+
+    osEventFlagsWait(networking_event, NETIF_CONFIGURED, osFlagsWaitAny, osWaitForever);
+
+    // Bind the socket to the multicast address and port we then use that comm_profile
+    // to join the specified multicast group.
+    struct netconn* conn = netconn_new(NETCONN_UDP_IPV6);
+
+    netconn_set_ipv6only(conn, true);
+    netconn_set_recvtimeout(conn, 1000);
+
+    netconn_bind(conn, &comm_profile->multicast_address, comm_profile->port);
+    netconn_join_leave_group(conn, &comm_profile->multicast_address, IP6_ADDR_ANY,
+                             NETCONN_JOIN);
+
+    struct netbuf* rx_buf = NULL;
+    err_t network_err;
+    int protobuf_err;
+
+    for (;;)
+    {
+        network_err = netconn_recv(conn, &rx_buf);
+
+        switch (network_err)
+        {
+            case ERR_TIMEOUT:
+            {
+                io_proto_multicast_communication_profile_notifyEvents(comm_profile,
+                                                                      RECEIVE_TIMEOUT);
+                break;
+            }
+
+            case ERR_OK:
+            {
+                pb_istream_t in_stream = pb_istream_from_buffer(
+                    (uint8_t*)rx_buf->p->payload, rx_buf->p->tot_len);
+
+                // deserialize into buffer
+                io_proto_multicast_communication_profile_acquireLock(comm_profile);
+                protobuf_err = pb_decode(&in_stream, comm_profile->message_fields,
+                                         comm_profile->protobuf_struct);
+                io_proto_multicast_communication_profile_releaseLock(comm_profile);
+
+                // nanopb err logic is inverted, false = error
```

Comment:
maybe renamed `protobuf_err` to `no_protobuf_err` or something to help readability as well?

Reply:
fixed

Reply:
nice, reads much better. nit: can we move this comment to the `pb_decode` line since I think that's where it's more relevant (the source of the nanopb error)

Reply:
fixed